tsl/src/compression/compress_utils.c

```diff
@@ -334,13 +342,22 @@ decompress_chunk_impl(Oid uncompressed_hypertable_relid, Oid uncompressed_chunk_
 					AccessShareLock);
 	LockRelationOid(catalog_get_table_id(ts_catalog_get(), CHUNK), RowExclusiveLock);
 
-	decompress_chunk(compressed_chunk->table_id, uncompressed_chunk->table_id);
+	decompress_chunk(compressed_chunk->table_id,
+					 uncompressed_chunk->table_id,
+					 reindex_uncompressed_chunk);
 	/* Recreate FK constraints, since they were dropped during compression. */
 	ts_chunk_create_fks(uncompressed_chunk);
 	ts_compression_chunk_size_delete(uncompressed_chunk->fd.id);
 	ts_chunk_clear_compressed_chunk(uncompressed_chunk);
-	ts_chunk_drop(compressed_chunk, DROP_RESTRICT, -1);
-	/* reenable autovacuum if necessary */
+
+	/* We only take an exclusive lock on the uncompressed chunk to prevent a
+	   deadlock if we are not going to reindex it. This is only when
+	   recompressing a chunk */
```

Comment:
If this really is only when recompressing then it should be:
```
decompress_chunk_impl(Oid uncompressed_hypertable_relid, Oid uncompressed_chunk_relid,
					  bool if_compressed, bool recompressing_chunk)
...
	decompress_chunk(compressed_chunk->table_id,
					 uncompressed_chunk->table_id,
					 recompressing_chunk ? /* reindex_out_table */ false : true);
...
	if (recompressing_chunk)
		LockRelationOid(uncompressed_chunk->table_id, AccessExclusiveLock);
```
and keep the rest:
```
decompress_chunk(Oid in_table, Oid out_table, bool reindex_out_table)
```

Reply:
That seems to be in the wrong order. The recompress inside `decompress_chunk` will take a lock on the index, but the order should be: relation first, index after.

Reply:
I'm not suggesting any change in order, I'm just renaming `reindex_out_table` to `recompressing_chunk` because it's very confusing as it currently is.

Reply:
This way we don't need the comment either.